pkg/aws/eni/node_manager.go

```diff
@@ -71,31 +73,46 @@ type NodeManager struct {
 	metricsAPI      metricsAPI
 	resyncTrigger   *trigger.Trigger
 	deficitResolver *trigger.Trigger
+	parallelWorkers int64
 }
 
 // NewNodeManager returns a new NodeManager
-func NewNodeManager(instancesAPI nodeManagerAPI, ec2API ec2API, k8sAPI k8sAPI, metrics metricsAPI) (*NodeManager, error) {
+func NewNodeManager(instancesAPI nodeManagerAPI, ec2API ec2API, k8sAPI k8sAPI, metrics metricsAPI, parallelWorkers int64) (*NodeManager, error) {
+	if parallelWorkers < 1 {
+		parallelWorkers = 1
+	}
+
 	mngr := &NodeManager{
-		nodes:        nodeMap{},
-		instancesAPI: instancesAPI,
-		ec2API:       ec2API,
-		k8sAPI:       k8sAPI,
-		metricsAPI:   metrics,
+		nodes:           nodeMap{},
+		instancesAPI:    instancesAPI,
+		ec2API:          ec2API,
+		k8sAPI:          k8sAPI,
+		metricsAPI:      metrics,
+		parallelWorkers: parallelWorkers,
 	}
 
 	deficitResolver, err := trigger.NewTrigger(trigger.Parameters{
 		Name:        "eni-node-manager-deficit-resolver",
 		MinInterval: time.Second,
 		TriggerFunc: func(reasons []string) {
+			sem := semaphore.NewWeighted(parallelWorkers)
+
 			for _, name := range reasons {
-				if node := mngr.Get(name); node != nil {
-					if err := node.ResolveIPDeficit(); err != nil {
-						node.logger().WithError(err).Warning("Unable to resolve IP deficit of node")
+				sem.Acquire(context.TODO(), 1)
+				go func(name string) {
```

Comment:
Why don't you have a pool of workers to that run in a N go routines where `reasons` is a channel? Seems cleaner than creating a new semaphore on each trigger.

```golang
for i :=0; i < parallelWorkers; i++ {
    go func (){
            for name := reasons {
            ...
            }
    ()
}
```

Reply:
I don't understand, what is cleaner?

Reply:
Having the same N of go routines instead of creating new ones each time and instead of creating a new semaphore. It is just a suggestion I'm fine either way.